// @chaosity/location-client/dist/auth/TokenProvider.js
export class TokenProvider {
    constructor(config) {
        this.config = config;
    }
    async getToken(forceRefresh = false) {
        if (!forceRefresh && this.cachedToken && this.expiresAt && !this.isExpired()) {
            return {
                success: true,
                token: this.cachedToken,
                expiresAt: this.expiresAt
            };
        }
        try {
            const response = await fetch(`${this.config.apiUrl}/auth/token`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    client_id: this.config.clientId,
                    client_secret: this.config.clientSecret,
                    grant_type: 'client_credentials',
                }),
            });
            if (!response.ok) {
                throw new Error(`Failed to get token: ${response.statusText}`);
            }
            const data = await response.json();
            this.cachedToken = data.access_token;
            this.expiresAt = Date.now() + (data.expires_in * 1000);
            return {
                success: true,
                token: this.cachedToken,
                expiresIn: data.expires_in,
                expiresAt: this.expiresAt,
            };
        }
        catch (error) {
            return {
                success: false,
                error: error instanceof Error ? error.message : 'Failed to get token',
            };
        }
    }
    isExpired(bufferSeconds = 60) {
        if (!this.expiresAt)
            return true;
        return Date.now() >= (this.expiresAt - bufferSeconds * 1000);
    }
    clearCache() {
        this.cachedToken = undefined;
        this.expiresAt = undefined;
    }
}
