// @chaosity/location-client/dist/adapters/GeoPlaces.js
import { placeToFeatureCollection } from '@aws/amazon-location-utilities-datatypes';
import { GetPlaceCommand, ReverseGeocodeCommand, SearchNearbyCommand, SuggestCommand } from '@aws-sdk/client-geo-places';
/**
 * GeoPlaces - MapLibre adapter for AWS Location Service
 *
 * Uses AWS Location Client commands and utilities for data conversion.
 * Only auth is custom (Bearer token instead of SigV4).
 */
export class GeoPlaces {
    constructor(client, map) {
        this.client = client;
        this.map = map;
    }
    async forwardGeocode(config) {
        const center = this.map.getCenter();
        const command = new SuggestCommand({
            QueryText: config.query,
            BiasPosition: [center.lng, center.lat],
            MaxResults: config.limit || 5,
            Language: 'en'
        });
        const response = await this.client.send(command);
        return placeToFeatureCollection(response, { flattenProperties: true });
    }
    async reverseGeocode(config) {
        const command = config.click
            ? new ReverseGeocodeCommand({
                QueryPosition: config.query,
                MaxResults: config.limit || 1,
                Language: 'en'
            })
            : new SearchNearbyCommand({
                QueryPosition: config.query,
                MaxResults: config.limit || 15,
                Language: 'en'
            });
        const response = await this.client.send(command);
        return placeToFeatureCollection(response, { flattenProperties: true });
    }
    async getSuggestions(config) {
        const center = this.map.getCenter();
        const command = new SuggestCommand({
            QueryText: config.query,
            BiasPosition: [center.lng, center.lat],
            Language: 'en'
        });
        const response = await this.client.send(command);
        return placeToFeatureCollection(response, { flattenProperties: true });
    }
    async searchByPlaceId(placeId) {
        const command = new GetPlaceCommand({
            PlaceId: placeId,
            Language: 'en'
        });
        const response = await this.client.send(command);
        return placeToFeatureCollection(response, { flattenProperties: true });
    }
}
