// @chaosity/location-client/dist/client/GeoPlacesClient.js
import { AutocompleteCommand, GeocodeCommand, GetPlaceCommand, ReverseGeocodeCommand, SearchNearbyCommand, SearchTextCommand, SuggestCommand, } from '@aws-sdk/client-geo-places';
/**
 * GeoPlacesClient - AWS Location Service compatible client with custom auth
 *
 * Uses AWS Location Client command classes but replaces SigV4 auth with Bearer token.
 * All request/response types are identical to AWS Location Service.
 */
export class GeoPlacesClient {
    constructor(config) {
        this.clientConfig = config;
        this.config = { serviceId: 'Geo Places' };
    }
    async send(command) {
        const endpoint = this.getEndpoint(command);
        const response = await fetch(`${this.clientConfig.apiUrl}${endpoint}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${this.clientConfig.token}`
            },
            body: JSON.stringify(command)
        });
        if (!response.ok) {
            throw new Error(`API request failed: ${response.statusText}`);
        }
        return response.json();
    }
    getEndpoint(command) {
        if (command instanceof AutocompleteCommand) {
            return '/address/autocomplete';
        }
        else if (command instanceof GeocodeCommand) {
            return '/address/geocode';
        }
        else if (command instanceof GetPlaceCommand) {
            return '/address/place';
        }
        else if (command instanceof ReverseGeocodeCommand) {
            return '/address/search/reverse-geocode';
        }
        else if (command instanceof SearchNearbyCommand) {
            return '/address/search/nearby';
        }
        else if (command instanceof SearchTextCommand) {
            return '/address/search/text';
        }
        else if (command instanceof SuggestCommand) {
            return '/address/suggestion';
        }
        throw new Error('Unknown command type');
    }
}
