// @aws/polyline/dist/cjs/polyline-types.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThirdDimension = exports.CompressionAlgorithm = exports.FlexiblePolylineFormatVersion = exports.DefaultPrecision = void 0;
/** Defines the default encoding precision for coordinates */
exports.DefaultPrecision = 6;
/** The version of flexible-polyline that's supported by this implementation */
exports.FlexiblePolylineFormatVersion = 1;
/** Defines the set of compression algorithms that are supported by this library. */
var CompressionAlgorithm;
(function (CompressionAlgorithm) {
    /** Encoder/decoder for the [Flexible Polyline](https://github.com/heremaps/flexible-polyline) format. */
    CompressionAlgorithm[CompressionAlgorithm["FlexiblePolyline"] = 0] = "FlexiblePolyline";
    /** Encoder/decoder for the [Encoded Polyline Algorithm Format](https://developers.google.com/maps/documentation/utilities/polylinealgorithm)
     * with 5 bits of precision.
     */
    CompressionAlgorithm[CompressionAlgorithm["Polyline5"] = 1] = "Polyline5";
    /** Encoder/decoder for the [Encoded Polyline Algorithm Format](https://developers.google.com/maps/documentation/utilities/polylinealgorithm)
     * with 6 bits of precision.
     */
    CompressionAlgorithm[CompressionAlgorithm["Polyline6"] = 2] = "Polyline6";
})(CompressionAlgorithm || (exports.CompressionAlgorithm = CompressionAlgorithm = {}));
/** Defines how to interpret a third dimension value if it exists. */
var ThirdDimension;
(function (ThirdDimension) {
    /** No third dimension specified */
    ThirdDimension[ThirdDimension["None"] = 0] = "None";
    /** Third dimension is level */
    ThirdDimension[ThirdDimension["Level"] = 1] = "Level";
    /** Third dimension is altitude (height above the Earth's surface) */
    ThirdDimension[ThirdDimension["Altitude"] = 2] = "Altitude";
    /** Third dimension is elevation (height of the Earth's surface relative to the reference geoid) */
    ThirdDimension[ThirdDimension["Elevation"] = 3] = "Elevation";
})(ThirdDimension || (exports.ThirdDimension = ThirdDimension = {}));
