// @aws-sdk/client-geo-places/dist-es/models/enums.js
export const TypePlacement = {
    AFTER_BASE_NAME: "AfterBaseName",
    BEFORE_BASE_NAME: "BeforeBaseName",
};
export const AutocompleteAdditionalFeature = {
    CORE: "Core",
};
export const AutocompleteFilterPlaceType = {
    LOCALITY: "Locality",
    POSTAL_CODE: "PostalCode",
};
export const AutocompleteIntendedUse = {
    SINGLE_USE: "SingleUse",
};
export const PostalCodeMode = {
    ENUMERATE_SPANNED_LOCALITIES: "EnumerateSpannedLocalities",
    MERGE_ALL_SPANNED_LOCALITIES: "MergeAllSpannedLocalities",
};
export const PlaceType = {
    BLOCK: "Block",
    COUNTRY: "Country",
    DISTRICT: "District",
    INTERPOLATED_ADDRESS: "InterpolatedAddress",
    INTERSECTION: "Intersection",
    LOCALITY: "Locality",
    POINT_ADDRESS: "PointAddress",
    POINT_OF_INTEREST: "PointOfInterest",
    POSTAL_CODE: "PostalCode",
    REGION: "Region",
    SECONDARY_ADDRESS: "SecondaryAddress",
    STREET: "Street",
    SUB_BLOCK: "SubBlock",
    SUB_DISTRICT: "SubDistrict",
    SUB_REGION: "SubRegion",
};
export const ValidationExceptionReason = {
    CANNOT_PARSE: "CannotParse",
    FIELD_VALIDATION_FAILED: "FieldValidationFailed",
    MISSING: "Missing",
    OTHER: "Other",
    UNKNOWN_FIELD: "UnknownField",
    UNKNOWN_OPERATION: "UnknownOperation",
};
export const GeocodeAdditionalFeature = {
    ACCESS: "Access",
    INTERSECTIONS: "Intersections",
    SECONDARY_ADDRESSES: "SecondaryAddresses",
    TIME_ZONE: "TimeZone",
};
export const GeocodeFilterPlaceType = {
    INTERPOLATED_ADDRESS: "InterpolatedAddress",
    INTERSECTION: "Intersection",
    LOCALITY: "Locality",
    POINT_ADDRESS: "PointAddress",
    POSTAL_CODE: "PostalCode",
    STREET: "Street",
};
export const GeocodeIntendedUse = {
    SINGLE_USE: "SingleUse",
    STORAGE: "Storage",
};
export const PostalAuthority = {
    USPS: "Usps",
};
export const PostalCodeType = {
    USPS_ZIP: "UspsZip",
    USPS_ZIP_PLUS_4: "UspsZipPlus4",
};
export const ZipClassificationCode = {
    MILITARY: "Military",
    POST_OFFICE_BOXES: "PostOfficeBoxes",
    UNIQUE: "Unique",
};
export const RecordTypeCode = {
    FIRM: "Firm",
    GENERAL: "General",
    HIGH_RISE: "HighRise",
    POST_OFFICE_BOX: "PostOfficeBox",
    RURAL: "Rural",
    STREET: "Street",
};
export const GetPlaceAdditionalFeature = {
    ACCESS: "Access",
    CONTACT: "Contact",
    PHONEMES: "Phonemes",
    SECONDARY_ADDRESSES: "SecondaryAddresses",
    TIME_ZONE: "TimeZone",
};
export const GetPlaceIntendedUse = {
    SINGLE_USE: "SingleUse",
    STORAGE: "Storage",
};
export const ReverseGeocodeAdditionalFeature = {
    ACCESS: "Access",
    INTERSECTIONS: "Intersections",
    TIME_ZONE: "TimeZone",
};
export const ReverseGeocodeFilterPlaceType = {
    INTERPOLATED_ADDRESS: "InterpolatedAddress",
    INTERSECTION: "Intersection",
    LOCALITY: "Locality",
    POINT_ADDRESS: "PointAddress",
    STREET: "Street",
};
export const ReverseGeocodeIntendedUse = {
    SINGLE_USE: "SingleUse",
    STORAGE: "Storage",
};
export const SearchNearbyAdditionalFeature = {
    ACCESS: "Access",
    CONTACT: "Contact",
    PHONEMES: "Phonemes",
    TIME_ZONE: "TimeZone",
};
export const SearchNearbyIntendedUse = {
    SINGLE_USE: "SingleUse",
    STORAGE: "Storage",
};
export const SearchTextAdditionalFeature = {
    ACCESS: "Access",
    CONTACT: "Contact",
    PHONEMES: "Phonemes",
    TIME_ZONE: "TimeZone",
};
export const SearchTextIntendedUse = {
    SINGLE_USE: "SingleUse",
    STORAGE: "Storage",
};
export const SuggestAdditionalFeature = {
    ACCESS: "Access",
    CORE: "Core",
    PHONEMES: "Phonemes",
    TIME_ZONE: "TimeZone",
};
export const SuggestIntendedUse = {
    SINGLE_USE: "SingleUse",
};
export const QueryType = {
    BUSINESS_CHAIN: "BusinessChain",
    CATEGORY: "Category",
};
export const SuggestResultItemType = {
    PLACE: "Place",
    QUERY: "Query",
};
